extensions/amp-form/0.1/amp-form.js

```diff
@@ -488,14 +493,59 @@ export class AmpForm {
       dev().error(TAG, 'Form submit service failed: %s', e);
     }
 
-    const varSubsFields = this.getVarSubsFields_();
-    if (this.xhrAction_) {
-      this.handleXhrSubmit_(varSubsFields, trust);
-    } else if (this.method_ == 'POST') {
-      this.handleNonXhrPost_();
-    } else if (this.method_ == 'GET') {
-      this.handleNonXhrGet_(varSubsFields);
+    // Get our special fields
+    const varSubsFields =
+      this.getVarSubsFields_();
+    const asyncInputs =
+      this.form_.getElementsByClassName(AsyncInputClasses.ASYNC_INPUT);
+
+    // Do any assertions we may need to do
+    // For NonXhrGET
+    // That way we can determine if
+    // we can submit synchronously
+    if (!this.xhrAction_ && this.method_ == 'GET') {
+      this.assertSsrTemplate_(false, 'Non-XHR GETs not supported.');
+      this.assertNoSensitiveFields_();
+
+      // If we have no async inputs, we can just submit synchronously
+      if (asyncInputs.length === 0) {
+
+        for (let i = 0; i < varSubsFields.length; i++) {
+          this.urlReplacement_.expandInputValueSync(varSubsFields[i]);
+        }
+
+        this.handleNonXhrGet_();
+        return;
+      }
+
+      event.preventDefault();
     }
+
+    // Set ourselves to the SUBMITTING State
+    this.setState_(FormState.SUBMITTING);
+
+    // Promises to run before submitting the form
+    const presubmitPromises = [];
+    presubmitPromises.push(this.doVarSubs_(varSubsFields));
+    iterateCursor(asyncInputs, asyncInput => {
+      presubmitPromises.push(this.getValueForAsyncInput_(asyncInput));
+    });
+
+    Promise.all(presubmitPromises).then(() => {
+      if (this.xhrAction_) {
+        this.handleXhrSubmit_(trust);
+      } else if (this.method_ == 'POST') {
+        this.handleNonXhrPost_();
+      } else if (this.method_ == 'GET') {
+        this.handleNonXhrGet_();
+      }
+    }).catch(error => {
```

Comment:
Added error handling here.

Reply:
This should be unnecessary: the error is already handled by `handleXhrSubmitFailure_`. The non-xhr cases don't need to be handled by AMP since the browser redirects in the non-xhr cases.

Reply:
Consider splitting the `then`/`catch` clauses here into two functions, `handlePresubmitSuccess_` and `handlePresubmitFailure_` and call them in the first and second arguments to the `then`. The difference between `promise.then(successFn).catch(errorFn)` and `promise.then(successFn, errorFn)` and is that the `catch` block in the first option will receive errors from the `then` or from the root promise, where the second option will only receive errors from the promise.

```js
Promise.all(presubmitPromises).then(
    () => this.handlePresubmitSuccess_(trust),
    error => this.handlePresubmitError_(error);
```

`handleXhrSubmit_` already handles its own errors, so something like this makes it clearer that the error handling code is only designed to respond to the presubmit promises, and not any of the methods in the success block

Reply:
Sounds great, thanks for the detailed explanation! 😄